wdl-lint/src/v1/matching_parameter_meta.rs

```diff
@@ -0,0 +1,180 @@
+//! A lint rule for matching parameter metadata.
+
+use std::collections::HashMap;
+
+use wdl_ast::experimental::v1::InputSection;
+use wdl_ast::experimental::v1::ParameterMetadataSection;
+use wdl_ast::experimental::v1::TaskDefinition;
+use wdl_ast::experimental::v1::TaskOrWorkflow;
+use wdl_ast::experimental::v1::Visitor;
+use wdl_ast::experimental::v1::WorkflowDefinition;
+use wdl_ast::experimental::AstToken;
+use wdl_ast::experimental::Diagnostic;
+use wdl_ast::experimental::Diagnostics;
+use wdl_ast::experimental::Span;
+use wdl_ast::experimental::VisitReason;
+
+use super::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// The identifier for the matching parameter meta rule.
+const ID: &str = "MatchingParameterMeta";
+
+/// Creates a "missing param meta" diagnostic.
+fn missing_param_meta(parent: &TaskOrWorkflow, missing: &str, span: Span) -> Diagnostic {
+    let (context, parent) = match parent {
+        TaskOrWorkflow::Task(t) => ("task", t.name()),
+        TaskOrWorkflow::Workflow(w) => ("workflow", w.name()),
+    };
+
+    Diagnostic::warning(format!(
+        "{context} `{parent}` is missing a parameter metadata key for input `{missing}`",
+        parent = parent.as_str(),
+    ))
+    .with_rule(ID)
+    .with_label(
+        "this input does not have an entry in the parameter metadata section",
+        span,
+    )
+    .with_fix(format!(
+        "add a `{missing}` key to the `parameter_meta` section with a detailed description of the \
+         input.",
+    ))
+}
+
+/// Creates an "extra param meta" diagnostic.
+fn extra_param_meta(parent: &TaskOrWorkflow, extra: &str, span: Span) -> Diagnostic {
+    let (context, parent) = match parent {
+        TaskOrWorkflow::Task(t) => ("task", t.name()),
+        TaskOrWorkflow::Workflow(w) => ("workflow", w.name()),
+    };
+
+    Diagnostic::warning(format!(
```

Comment:
```suggestion
    Diagnostic::note(format!(
```
Per [#3](https://github.com/stjude-rust-labs/sprocket/issues/3) of sprocket, our team is actually out of step with idiomatic WDL. Current common practice is to document outputs in the parameter_meta. I think we should drop the level of "extra_param_meta" from `warning` to `note` because we are going against the grain with this lint.

In a future release we'll want configuration to permit `output` in parameter_meta as well as inputs (But I think we hold our position with a default behavior of warning on outputs)

Open to other ideas as well here. A bit of a tricky situation

Reply:
A note level makes sense to me.

Additionally, the spec does mention that both inputs and outputs should be present in the `parameter_meta` section and they _must_ correspond to an input or output name, but doesn't appear to put a restriction on a name that might appear as _both_ an input and an output; are users expected to document that the name is both an input and an output in the same entry, even if they might have different meanings?

Also, "parameter" definitely is a weird name for _results_ of a task or workflow. To an outsider coming into WDL, `parameter_meta` seems bolted on to the language. I hope that a future version can be steered towards documentation comments or at least separate `parameter_meta` into two sections: one for parameters/inputs and one for results/outputs.

Reply:
> but doesn't appear to put a restriction on a name that might appear as both an input and an output

Inputs and outputs are defined in the same namespace, so this case actually is covered. You can't define an output with the same name as an input.

> Also, "parameter" definitely is a weird name for results of a task or workflow. To an outsider coming into WDL, parameter_meta seems bolted on to the language. I hope that a future version can be steered towards documentation comments or at least separate parameter_meta into two sections: one for parameters/inputs and one for results/outputs.

As to this, I think our whole team is in agreement on that. However, our team appears to be a minority. I've tried to register a formal complaint [here](https://github.com/openwdl/wdl/issues/637) that has been largely ignored.

Reply:
> Inputs and outputs are defined in the same namespace, so this case actually is covered. You can't define an output with the same name as an input.

Good point, I had forgotten about that as we're not at a stage yet where we'd be evaluating the WDL to catch such errors.

> As to this, I think our whole team is in agreement on that. However, our team appears to be a minority. I've tried to register a formal complaint https://github.com/openwdl/wdl/issues/637 that has been largely ignored.

I'm happy to learn I'm not alone in my opinion 😅